[VIP-1939] Display integrations in plugins list in WP-Admin

shekharnwagh opened the pull request:
<!--
## For Automatticians!

:wave: Just a quick reminder that this is a public repo. Please don't include any internal links or sensitive data (like PII, private code, customer names, site URLs, etc. Any fixes related to security should be discussed with Platform before opening a PR. If you're not sure if something is safe to share, please just ask!

## For external contributors!

Welcome! We look forward to your contribution! ❤️
-->
## Description
<!--
A few sentences providing more context and describing the overall goals of the Pull Request.

Should include any special considerations, decisions, and links to relevant publicly available sources (e.g. GitHub issues)
-->

This PR adds functionality to display plugins loaded by VIP Integrations in the WordPress admin plugins list, providing better visibility.

Plugins loaded by VIP Integrations are loaded via `require_once` from the `mu-plugins/vip-integrations/` directory. Since WordPress's `get_plugins()` only scans the standard `wp-content/plugins/` directory, these integration plugins are invisible in the wp-admin plugins list.

To solve this, created a reusable `IntegrationPluginDisplayTrait` that:
- Injects integration plugins into WordPress's plugins list via the `all_plugins` filter
- Shows plugins as active with "Enabled via WPVIP Integrations" status
- Removes activate/deactivate links (like code-activated plugins)
- Prevents accidental database writes for integration plugins
- Suppresses "Plugin file does not exist" error messages using output buffering

This approach is similar to how `wpcom_vip_load_plugin()` displays code-activated plugins, but tweaked for the plugin code being in `mu-plugins/vip-integrations` dir.

<img width="1435" height="807" alt="Screenshot 2025-10-03 at 12 17 33 AM" src="https://github.com/user-attachments/assets/308451c4-f243-478c-930e-54f3a9e446f6" />


## Changelog Description

<!-- Changelogs are published for our customers. Well-written entries help them stay informed on platform changes and all of the great work that we do! -->

<!-- Write a concise description of changes in the relevant section.
- Add new line items as needed.
- Entries should follow the [Common Changelog Style Guide](https://github.com/vweevers/common-changelog). 
- Remove all unused sections before merging.
- Proof-read.
-->

### Added
- VIP Integrations: Added a reusable `IntegrationPluginDisplayTrait` that displays integration plugins in the WordPress admin plugins list
- VIP Integrations: When Real-Time Collaboration integration is loaded, the custom build of Gutenberg and the latest version of the Real-Time Collaboration plugin activated via VIP Integrations are displayed in the WordPress admin plugins list


## Pre-review checklist

Please make sure the items below have been covered before requesting a review:

- [ ] This change works and has been tested locally or in Codespaces (or has an appropriate fallback).
- [ ] This change works and has been tested on a sandbox.
- [ ] This change has relevant unit tests (if applicable).
- [ ] This change uses a rollout method to ease with deployment (if applicable - especially for large scale actions that require writes).
- [ ] This change has relevant documentation additions / updates (if applicable).
- [ ] I've created a changelog description that aligns with the provided examples.

## Pre-deploy checklist

- [ ] VIP staff: Ensure any alerts added/updated conform to internal standards (see internal documentation). 

## Steps to Test

1. Check out PR.
2. Copy the `/vip-integrations` directory from the [vip-go-mu-plugins-built](https://github.com/Automattic/vip-go-mu-plugins-built) repo to `/vip-integrations` directory of this repo.
3. Run `composer install` and `npm install` to install dependencies.
4. Activate the 'Real-Time Collaboration' integration on one of your test sites and note down its application ID.
5. Spin up a local dev for that site using command - `vip dev-env create @<application_id>.production -s <slug> -a <path_to_app_code> -u <path_to_local_vip-go-mu-plugins_dir>`
6. Start it using the command - `vip dev-env start --slug <slug> --editor vscode`
7. That should open a VS Code window. Modify the `<app_name>/integrations-config/integrations.json` to add following entry -

	```json
	"real-time-collaboration": {
		"type": "real-time-collaboration",
		"org": {
			"status": "enabled"
		},
		"env": {
			"status": "enabled",
			"config": {
				"web_socket_auth_secret": "test-secret-key",
				"web_socket_url": "wss://test.example.com/_ws"
			}
		}
	}
	```
8. Navigate to the `wp-admin` > `Plugins` page and verify that the custom build of Gutenberg and the latest version of the Real-Time Collaboration plugin activated via VIP Integrations are displayed in the WordPress admin plugins list.

shekharnwagh committed:
add RTC and gutenberg entry in plugins list when integration is activated

shekharnwagh committed:
add IntegrationPluginDisplayTrait to manage integration plugin display

shekharnwagh committed:
hard code the integration plugin message

shekharnwagh committed:
use static keyword instead of self

shekharnwagh committed:
update comments

shekharnwagh committed:
register RTC integration plugins only after successful load

Previously, Gutenberg and RTC plugins were registered for display
immediately after require_once in RealTimeCollaborationIntegration,
which meant they could appear in the WordPress plugins list even if
they failed to load due to compatibility checks or other issues.

shekharnwagh committed:
add tests for IntegrationPluginDisplayTrait

shekharnwagh committed:
refactor tests to use process isolation

The trait uses static properties shared across all integrations.
Previous approach used a manual reset method between tests which
added test-specific code to trait.

shekharnwagh committed:
fix overly greedy regex in error notice suppression

The previous pattern used /s modifier and .*? which could match
across multiple error divs, causing it to remove unrelated error
notices.

pkevan commented on integrations/integration-plugin-display-trait.php:
```diff
@@ -0,0 +1,386 @@
+	 *
+	 * @var string
+	 */
+	private static $plugin_display_message = 'Enabled via WPVIP Integrations';
```
is there something more concise we can use here - feels a little long, particularly when it spans two lines.

shekharnwagh replied:
Open to ideas. Worded it so it's similar to badge - `Enabled via code` shown when we [activate plugin from code](https://docs.wpvip.com/plugins/activate-plugins-through-code/) using `wpcom_vip_load_plugin`.

What do we think using just `WPVIP Integration` ? But that might be slightly confusing when that shows up for `Gutenberg` plugin.

pkevan replied:
Potentially something as simple as a link to the integrations center via a link `Settings`?

pkevan replied:
![Screenshot 2025-10-10 at 09 09 46](https://github.com/user-attachments/assets/5380c963-a2f3-4b6b-9b1a-901c442cc18e)

shekharnwagh replied:
Just took a quick look at the code in this repo and I couldn't find any references to `dashboard.wpvip.com`. If we have add it, it'll be some work constructing the url and maintaining it in unlikely scenario the routing in dashboard changes.

pkevan replied:
perhaps we get away with a link to the docs? like: https://docs.wpvip.com/integrations/

ingeniumed replied:
How about `Enabled via integrations` and integrations links to https://docs.wpvip.com/integrations/?

maxschmeling replied:
I wouldn't be against adding a link or something better down the road, but the simple but clear text that you already have in there is fine with me even though it splits to two lines.

rinatkhaziev reviewed (changes requested):
So using regex and ob doesn't seem entirely safe. There are better ways, for example:

https://developer.wordpress.org/reference/hooks/plugin_action_links/

pkevan commented on integrations/integration-plugin-display-trait.php:
```diff
@@ -0,0 +1,386 @@
+
+		// Cleanup and error suppression
+		add_action( 'admin_init', [ $this, 'cleanup_recently_activated' ], 1 );
+		add_action( 'admin_notices', [ $this, 'suppress_deactivation_notices' ], 1 );
```
curious as to when these get triggered, as it looks like they would only appear when the plugin activation/deactivation links are present and a user clicks on them?

shekharnwagh replied:
Thinking through this again, yeah we won't need this since we've already removed the links to activate/deactivate plugin and the activation happens via code. Removed in a38323ebf

maxschmeling commented on integrations/integration-plugin-display-trait.php:
```diff
@@ -0,0 +1,386 @@
+		static::$buffer_started = true;
+
+		// Use output buffering to filter out error notices about integration plugins
+		ob_start( function ( $buffer ) use ( $integration_plugins ) {
```
```suggestion
		// Use output buffering to filter out error notices about integration plugins
		static::$buffer_started = ob_start( function ( $buffer ) use ( $integration_plugins ) {
```

maxschmeling committed:
Merge branch 'develop' into add/integration-plugins-entry

shekharnwagh committed:
Merge branch 'develop' into add/integration-plugins-entry

shekharnwagh committed:
Remove unused cleanup_recently_activated hook

Integration plugins never enter the recently_activated list since
they're filtered out of active_plugins and users can't deactivate
them via the UI. This method is therefore unnecessary.

shekharnwagh commented:
> So using regex and ob doesn't seem entirely safe. There are better ways, for example:
> 
> https://developer.wordpress.org/reference/hooks/plugin_action_links/

The `plugin_action_links` only customizes the action buttons in each plugin row. The "Plugin file does not exist" error is generated during WordPress's plugin validation. We are already hooking into that filer to remove the activation and deactivation links and adding the `Enabled via WPVIP Integrations` message.
That's why I ended up using output buffering to suppress the error notice.

I'm open to better solutions. Is there a specific WordPress hook or filter that prevents the validation check for plugins in non-standard directories? @rinatkhaziev

shekharnwagh committed:
use return value of ob_start

Co-authored-by: Max Schmeling <max.schmeling@automattic.com>

shekharnwagh committed:
Merge branch 'develop' into add/integration-plugins-entry

rinatkhaziev commented:
> I'm open to better solutions. Is there a specific WordPress hook or filter that prevents the validation check for plugins in non-standard directories? @rinatkhaziev

OIC, I misunderstood the diff.

Okay, in this particular case simply removing `filter_active_plugins_for_display` gets rid of the notice.

```diff
diff --git a/integrations/integration-plugin-display-trait.php b/integrations/integration-plugin-display-trait.php
index 9376cfb55..ce872a255 100644
--- a/integrations/integration-plugin-display-trait.php
+++ b/integrations/integration-plugin-display-trait.php
@@ -122,7 +122,7 @@ trait IntegrationPluginDisplayTrait {
 		add_filter( 'network_admin_plugin_action_links', [ $this, 'filter_plugin_action_links' ], 10, 2 );
 
 		// Active plugin management (display only)
-		add_filter( 'option_active_plugins', [ $this, 'filter_active_plugins_for_display' ] );
+		// add_filter( 'option_active_plugins', [ $this, 'filter_active_plugins_for_display' ] );
 		add_filter( 'site_option_active_sitewide_plugins', [ $this, 'filter_network_active_plugins_for_display' ] );
 
 		// Prevent accidental database writes
@@ -130,8 +130,8 @@ trait IntegrationPluginDisplayTrait {
 		add_filter( 'pre_update_site_option_active_sitewide_plugins', [ $this, 'filter_update_network_active_plugins' ] );
 
 		// Cleanup and error suppression
-		add_action( 'admin_notices', [ $this, 'suppress_deactivation_notices' ], 1 );
-		add_action( 'network_admin_notices', [ $this, 'suppress_deactivation_notices' ], 1 );
+		// add_action( 'admin_notices', [ $this, 'suppress_deactivation_notices' ], 1 );
+		// add_action( 'network_admin_notices', [ $this, 'suppress_deactivation_notices' ], 1 );
 	}
 
 	/**

```
<img width="2542" height="2076" alt="CleanShot 2025-10-24 at 18 13 53@2x" src="https://github.com/user-attachments/assets/c0415eac-2ba2-4cd0-9de4-c3616a4cdeba" />

shekharnwagh commented:
> Okay, in this particular case simply removing filter_active_plugins_for_display gets rid of the notice.

Removing `filter_active_plugins_for_display` means they now show up in inactive plugins list which isn't what we'd want. We do have a text saying "Enabled via WPVIP Integrations", but they not being in active list might indicate they aren't loaded.

<img width="1737" height="967" alt="Screenshot 2025-10-25 at 10 40 00 AM" src="https://github.com/user-attachments/assets/49be116c-2d01-46cc-8af4-862df8fdf953" />

shekharnwagh commented:
> > Okay, in this particular case simply removing filter_active_plugins_for_display gets rid of the notice.
> 
> Removing `filter_active_plugins_for_display` means they now show up in inactive plugins list which isn't what we'd want. We do have a text saying "Enabled via WPVIP Integrations", but they not being in active list might indicate they aren't loaded.
> <img alt="Screenshot 2025-10-25 at 10 40 00 AM" width="1737" height="967" src="https://private-user-images.githubusercontent.com/41412488/505584218-49be116c-2d01-46cc-8af4-862df8fdf953.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE4NDQxNjksIm5iZiI6MTc2MTg0Mzg2OSwicGF0aCI6Ii80MTQxMjQ4OC81MDU1ODQyMTgtNDliZTExNmMtMmQwMS00NmNjLThhZjQtODYyZGY4ZmRmOTUzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDMwVDE3MDQyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM1YzVlMTRlNDhhYTM5ZjIzOGI4YTJmZmMyYzliODY3ZDVlZmUxNDgwMjEyMmQ0Nzk4NTQ3M2ViOGE3YTBlNDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.G4VyKvx-UIoTNcndImwTRhhD8nIVKwTX_14ulII7nAY">

Bringing this on your radar in case it slipped through - @rinatkhaziev

rinatkhaziev commented:
Still very much on the radar, I asked @sjinks to look at it to provide ideas/suggestions. @shekharnwagh

sjinks commented:
I am exploring how to (ab)use the `plugins_list` filter: https://github.com/WordPress/WordPress/blob/master/wp-admin/includes/class-wp-plugins-list-table.php#L309

The idea is that we can add the "Integrations" tab to the plugins list and display our integrations there.

I hope that this approach is less fragile and does not require changes to integrations' code.

shekharnwagh commented:
> I am exploring how to (ab)use the `plugins_list` filter: https://github.com/WordPress/WordPress/blob/master/wp-admin/includes/class-wp-plugins-list-table.php#L309
> 
> The idea is that we can add the "Integrations" tab to the plugins list and display our integrations there.
> 
> I hope that this approach is less fragile and does not require changes to integrations' code.

@sjinks did your exploration of the `plugins_list` filter yield any results?

shekharnwagh committed:
Merge branch 'develop' into add/integration-plugins-entry
